Filter command: Add '--where' option for complex queries

danielsoneg opened the pull request:
Per discussion in #244 , adds a `--where` option to `augur filter` to perform complex queries by leveraging Pandas DataFrame.query() method

I'm new and mostly trying to work off some nervous energy while stuck indoors, so I defer to the users and maintainers on whether this is the right implementation. There's some issues with the approach:

Because of where the metadata filtering takes place, this winds up being a bit messy for a few reasons:
1. there's no way to tell later during filtering whether an entry is missing metadata because of the filter or because it's just missing metadata.
2. Because we're dropping the metadata before any of the rest of the filtering happens, `--include` and `--include-with` can't override the query.
3. Because we're just passing a string directly to DataFrame.query(), we can't compare dates or do substring filtering. We can only compare integers and exact strings. Finally, Pandas' error handling for query failures is nonspecific
4. I got a handful of different errors when trying to break this, some Pandas specific and some not, so the error message is effectively "You did a bad filter, here's what pandas said."

It's possible to improve 1 and 2 by reworking what `utils.read_metadata` returns, but that'd require updating everything else that uses read_metadata and I didn't want to step in that for my first patch - this is intentionally a minimal change.

All that said, this gets the job done as discussed in #244 and is probably worth it for the added functionality, but I'm not going to feel put out if y'all pass on this one.

Oh, and - example output, using the nextstrain/zika-tutorial files:
Good:
```
% augur filter \
  --sequences data/sequences.fasta \
  --metadata data/metadata.tsv \
  --min-date 2012 \
  --group-by country year month \
  --sequences-per-group 1 \
  --output filtered.fasta \
  --where="(journal == 'Nature (2017) In press') | (region == 'North America')"
No meta data for COL/FLR_00024/2015, excluding from all further analysis.
No meta data for COL/FLR_00008/2015, excluding from all further analysis.
<...snip...>
No meta data for SMGC_1, excluding from all further analysis.

24 sequences were dropped during filtering
	0 of these were dropped because of their date (or lack of date)
	3 of these were dropped because of subsampling criteria

	21 sequences were dropped because they did not have metadata. This may be due to the metadata query:
		(journal == 'Nature (2017) In press') | (region == 'North America')
10 sequences have been written out to filtered.fasta
```
Bad query:
```
% augur filter \
  --sequences data/sequences.fasta \
  --metadata data/metadata.tsv \
  --min-date 2012 \
  --group-by country year month \
  --sequences-per-group 1 \
  --output filtered.fasta \
  --where="(foo == bar)"
ERROR: Invalid query string: '(foo == bar)'
name 'foo' is not defined
```
Numeric:
```
(augur) [17:45] eric@tceskia:nextstrain/zika-tutorial ‹master*›
% augur filter \
  --sequences data/sequences.fasta \
  --metadata data/metadatawithnumbers.tsv \
  --min-date 2012 \
  --group-by country year month \
  --output filtered.fasta \
  --where="(length >= 1000)"
No meta data for BRA/2016/FC_6706, excluding from all further analysis.
No meta data for Aedes_aegypti/USA/2016/FL05, excluding from all further analysis.
<...snip...>
No meta data for Brazil/2015/ZBRC303, excluding from all further analysis.

9 sequences were dropped during filtering
	0 of these were dropped because of their date (or lack of date)

	9 sequences were dropped because they did not have metadata. This may be due to the metadata query:
		(length >= 1000)
25 sequences have been written out to filtered.fasta
```

rneher commented:
This is a nice and simple way to implement a generic filter. as you point out, it goes a bit against the grain of the rest as sequences are excluded for lack of metadata. But I think the simplicity and flexibility makes up for that.

emmahodcroft commented:
My only initial comment would be that we should change the message that'd displayed: 

> `9 sequences were dropped because they did not have metadata. This may be due to the metadata query:(length >= 1000)`

being dropped because of a query & because they don't have metadata are quite different. If you are missing metadata this isn't very helpful for tracking down where things are going wrong. I'd say we should change this to be more accurate - and add to the current format where if you're dropped it says something like `dropped due to --exclude-where blah = blah` (but modified for the new query format)

rneher commented:
@emmahodcroft If the output was the following:
```
3 sequences were dropped during filtering
        3 of these were dropped because the metadata prefilter:
                country=='Brazil' or country!='Colombia'
9 sequences have been written out to tmp.fasta
```
would that be helpful?

rneher commented:
I renamed the flag
 `--prefilter-metadata "country=='Brazil' or country!='Colombia'"`
and it seems to work as intended.

emmahodcroft commented:
But wait, it still can't tell the difference between missing metadata and prefiltered metadata?

emmahodcroft commented:
I'd rather read the metadata in twice if that's the case, to get the diff. I think alerting people to missing metadata is important. And telling people how many were dropped due to their filter is important. But they're really different problems.

rneher commented:
that is what I have. 
```
    num_excluded_lacking_metadata = 0
    num_excluded_metadata_filter = 0
    for seq_name in seq_keep:
        if seq_name in unfiltered_meta_dict:
            if seq_name in meta_dict:
                tmp.append(seq_name)
            else:
                num_excluded_metadata_filter += 1
        else:
            print("No meta data for %s, excluding from all further analysis."%seq_name)
            num_excluded_lacking_metadata +=1
    seq_keep = tmp
```

rneher commented:
commits:
https://github.com/nextstrain/augur/commit/b439eeedfa75798b4817f082211da67aaa34ff34
https://github.com/nextstrain/augur/commit/06ac42fdfad86be067be17919f4300c60807bfd0

emmahodcroft commented:
If the number excluded for not having metadata is separate from the number excluded because of a filter, then that is good to me!

danielsoneg commented:
@emmahodcroft @rneher I think the cleaner option is to update the read_metadata function to return the list of filtered strains and update any other place that's calling it. I'd considered that when I first wrote this, but it was my first PR for the project so I was keeping it minimal. There's only 6 different calls to the function, so it's not a huge change.

rneher commented:
the problem will be that this function is also used by various scripts in different analysis like here:

https://github.com/nextstrain/ncov/blob/master/scripts/construct-recency-from-submission-date.py#L34

these could be replaced by simple calls to pandas, but one would need to find them all...

danielsoneg commented:
oh my.

So there's a right way, a wrong way, and a Max Power way to handle this.

The right way is probably to refactor and split the function - `read_metadata` and `read_metadata_with_filter`, and re-use the internals of that function. 

The wrong way - and I _really_ don't like this option, but -
Python isn't strictly typed, we could just change the return signature if we got a query.

The Max Power way is what @emmahodcroft suggested, just read the metadata twice, once with the filter, once without. I didn't do that in the first place because I don't know the performance implications (except, obviously, that it's going to take twice as long to read the metadata 😁)

I'm partial to doing it the Right Way, because frankly I've got the time and I think it's worth keeping the code clean, but this is y'alls rodeo, so I'll go whichever way you prefer.

rneher commented:
agreed. This reuse of `utils` is a bit of a liability. The "right way" seems good to me. I already played with the "max power" way a bit and I'd be fine with that as well. reading the metadata is negligible compared to all other computations. I pushed my messing around to the `nextstrain/augur` repo just to be able to share:

https://github.com/nextstrain/augur/commit/b439eeedfa75798b4817f082211da67aaa34ff34#diff-96efef10f1bda965f758dd80e884f791L123

danielsoneg commented:
Ok, that took a bit of work because the error/return flow's a bit odd for read_metadata, but - 
I split the guts of `read_metadata` into two other functions: `load_metadata_files`, which handles the pandas bits, and `format_metadata`, which handles shifting the dataframe into a dictionary. I added another function, `read_metadata_with_query`, which does almost exactly what `read_metadata` does and then filters the results, returning the metadata dictionary and column list as well as a set of filtered strains. I then updated `filter.py` to use the new function and report how many strains were dropped:

```
% augur filter --metadata meta.tsv -s tb.vcf.gz -o test.vcf --prefilter-metadata="location=='other_village'"
No meta data for G22676, excluding from all further analysis.
Filtering samples using VCFTools with the call:
vcftools --remove-indv G22677 --remove-indv G22676 --remove-indv G22614 --remove-indv G22592 --remove-indv G22615 --remove-indv G22574 --remove-indv G22611 --remove-indv G22601 --remove-indv G22665 --gzvcf tb.vcf.gz --recode --stdout  > test.vcf

9 sequences were dropped during filtering
        1 of these were dropped because they did not have metadata

        8 were dropped due to the metadata query:
                location=='other_village'
1 sequences have been written out to test.vcf
```

danielsoneg commented:
(One quick note on this - this PR was originally opened ~2mo ago, which means there's a ton that's changed on master since this was done. I'm holding off on rebasing until we're all happy with the implementation, and I'm holding off on tests until after rebasing because there's been a ton of test files written since then (including for utils) and I don't want to create _more_ merge conflicts)

rneher commented:
thanks! this looks very nice and clean to me. 

@huddlej, why don't you have a look and then @danielsoneg can rebase? I don't expect much fuzz with testing since this shouldn't change the behavior when no `--prefilter-metadata` is specified. we might also want to have a last minute vote on the name of the flag. The `--prefilter-metadata` has the advantage that it makes it clear that this is something that is applied before everything else, including subsampling. but out of context it is a little cryptic. Also we already have `include`, `exclude`, `include-where` and `exclude-where` all with slightly different semantics. so another `where` seems exceedingly confusing.  

(and yes, sorry this has been sitting there for 2 months -- just trying to catch up now)

danielsoneg commented:
No worries on the time, you all have more important things to do right now than worry about some dude on the internet 😁

danielsoneg commented:
So I really like the idea of using Pandas for general querying - there's basically three things filter.py does:

1. Filter based on sequence factors - length and invalid nucleotides
2. Filter based on metadata and an include/exclude list
3. Subsample based on various criteria

I think reducing everything in #2 to a ~single function would be a big win for sure. I also think it's worth considering moving Subsample out to its own command - it's a fundamentally different process than the other two.

Date handling gets a bit weird in this case - there's some code in there for handling dates like "1991-XX-XX", which seem common, and there's no good way to cleanly handle that with the query language. I think Pandas can probably handle that better than the current methods, but it would still have to be a separate set of flags & code path.

The big challenge  is the migration path from the old version to the new version. I honestly think that, due to the differences between the two versions and the current state of filter.py, running the two code paths at the same time would be a nightmare - I think the only really tractable way to do this is either rename `filter` to `filter-legacy` or add a `--legacy` or some such flag to gate the behavior.

rneher commented:
I'd be happy with deprecating `exlude-where` and using `--query`. but I think the historical baggage doesn't allow and optimal solution...

emmahodcroft commented:
I think I agree with @huddlej  here. We need to keep this working 'the same' for people who are used to existing interface, without seriously changing how arguments work. The `where` or `query` arguments should just be an optional extra on top of this. 

I'd be open to thinking about a longer term depreciation of some options, but this wouldn't change anything immediately, and we'd need to ensure our docs are up-to-date for how to handle this. I'm not sure now is the time to do this (we are all super-stretched already!). I also think it's important we maintain some easy-to-use options for users who aren't so code-savvy.

danielsoneg commented:
@emmahodcroft Oh yeah, strong agree, especially on prioritization. I think there's a long-term path to a better filter, but I don't think now's the right time to grab that. 

I'm happy to help with anything that is high priority right now, though - I've got cycles (or at least will later this week).

huddlej commented:
Thanks all! I definitely don't think a full move to pandas is appropriate now, but I wanted to get @danielsoneg's perspective on it.

The short-term priority would be to merge this with `--query` and with support for running `--include` at the same time. I think this means we have to read in the metadata twice (once for the query samples and once for the include/exclude samples). Performance is not an issue here especially with the file sizes we are working with now.

@danielsoneg Would you be open to trying a modification of this current code to all both `--query` and `--include`?

danielsoneg commented:
Yep, I can do that - I’m busy through Friday, but I’ll take a crack at it over the weekend

emmahodcroft commented:
Fantastic @danielsoneg , thank you! And thank you @huddlej for doing the coordinating! Looking forward to having this new functionality, I think it'll open some exciting new paths :)

danielsoneg commented:
@huddlej @rneher @emmahodcroft I just submitted #555, which is effectively a rework of this pr against current master, with tests & updates per the conversations above. If that looks good, this can be closed.
